### DISCUSSION TITLE
Mojo and SEO

### DISCUSSION THREAD

**Discussion by lordfeck** [mojolicious/mojo#2001] [Q&A]
Greetings all,

I plan to write a web application and have been looking at Mojo and so far it seems perfect for my web development needs. 

However, SEO will be important to us. I'm not yet familiar with all that is needed for SEO but as far as I know it is down to getting naming right, meta tags and a sitemap in place. I'd definitely like the ability to generate a sitemap.xml, where the web routes are listed and ranked by importance for web crawlers. I'd also like to autogenerate the relevant meta tags to assist SEO and integration with social media platforms.

It is easy to append additional headers and I could even just write a helper function for the meta headers. But for the sitemap, is there already a Perl or Mojo plugin that automates this task?

Also, has anyone experiences of SEO and making big G happy with your Mojo site? All my web projects so far have been non-commercial, so this is an entirely new field for me.

Advice and stories will be appreciated.

Best, Lord Feck.

**Comment by ajmetz** [comment]
I too would be interested in hearing of any sitemap automation tools.  
I was likely to make one manually, lol.

I have a Catalyst app that I've also ported to Mojolicious.
Currently the Catalyst version is live (on a production server), and  
I'll switch it to the Mojolicious one in due course (that's currently  
running on a dev server).
And I really got the SEO down with just highly appropriate urls.

It's a (very) simple e-commerce shop selling videogame magazines,
and we're on the first page of google for most of the magazines if  
they are searched for with the issue number.
The urls are pretty much site name / magazine name - issue number, i.e.:
www.gamesmags.com/pokemon-73
However, www.gamesmags.com/pokemon will always show the current issue  
of pokemon, and because that page regularly updates, google views our  
site as a regularly updated or active site, and keeps updating its  
listings.
You will find us the third result for a search for "Pokemon Issue 73"  
on google.com - this is when using a private browser with no past  
cookies saved, so no personalised search results.

I currently don't have a sitemap, so google is not informed of the  
frequency with which the pokemon page is updated. So what I've found  
is that if I want people visiting my site on the day a new issue is  
released, I should go to Google WebMaster Tools, and inform google the  
page has updated, the day before release. Google will then have  
refreshed its data on my site, and when people search the next day for  
the latest issue, we'll be one of the first search results. As you can  
see - I've been lazy and not done this recently, so Google is showing  
a photo of issue 72 and text about issue 72, in the issue 73 search  
results, even though when you click through on the link, you get info  
on issue 73. So I need to go login in to Google WebMaster Tools  
now...except I think it had its name changed - shows how long it has  
been since I went on there - yeah - just logged in - it's called  
Google Search Console now.

So I understand having a relevant url, and having your page updated regularly.
I'm currently not doing anything with meta tags or site maps,
and I likely have far more to learn.

We are typically on the first page of search results when people  
search for any of the magazine titles. As our URL scheme is basically  
the magazine title. So our URLs are essentially the same as the search  
term.

Because my site currently runs at a small loss due to free postage, I  
haven't been keen to promote it until I get some startup funding to  
shift to higher volumes of stock I can make a profit on. Otherwise I  
would be caring a lot more about this, and trying to keep the site at  
the top of google more, and actively promote it. But yeah,  
search.google.com is your friend, =).

I also recommend Exposure Ninja -  
https://www.youtube.com/c/ExposureNinja/videos
ran by Tim - https://www.youtube.com/user/timlovesdrums/videos - who  
is an absolute expert on SEO and shares loads of good tips and advice,  
=).

Technically, my site should be in stealth/non-trading, however the SEO  
of the URLS means people always find us, and buy so I continue to lose  
money on free postage, =).
Oh well!




Quoting Lord Feck ***@***.***>:

**Comment by lordfeck** [comment]
@ajmetz thank you for the informative reply, this has given me a good few pointers. Thankfully we're not working in the same industry so I won't be stealing any of your steam ;)

I've also considered building a sitemap automation tool for Mojo - simply because I am lazy and I'd rather make a script than click something on a website regularly. I have used a similar tool with Pelican to automate sitemaps for my blog, so I'll look to how that was built to give me ideas.

If it is feasible I'll build this tool as part of our development roadmap and I'll let you know if it is finished. If it is good enough, I'll even offer it to CPAN as a Mojo plugin.

**Comment by guest20** [comment]
You can get your routing table from `Mojolicious::Routes`, the trick is figuring out how provide the values for all your placeholders. 

There might be some juice in [defining custom types](https://docs.mojolicious.org/Mojolicious/Routes#types) for all your placeholders and mapping those type names to particular queries on your content.

Obviously querying out all the content for your website is pretty costly, so you'd want to shove it in a cron job that just dumps it in `public/sitemap.xml` so it can be served statically. Maybe even keep track of the last time you wrote out your site map out so you can skip re-generating it if nothing has changed (based on create/update dates in your database). Or maybe write an item to a Minion queue to regenerate it when new stuff is published/updated...

**Comment by lordfeck** [comment]
@guest20 Thanks for the pointers, yeah now I see it'll be difficult to cover all the routes when we're just passing the placeholder to the DB lookup. I'd like to write this so it could be used as a generic Mojo plugin, so I'll investigate a few methods.

I can't promise an exact date but we've built quite a bit of our site in Mojo (enjoying the framework if I may say so) and a sitemap is needed. I should get around to it in the next few months
